File: course1_algorithmic_toolbox/week1_programming_challenges/2_maximum_pairwise_product/max_pairwise_product.js
const readline = require('readline');
const rl = readline.createInterface({
	input: process.stdin,
	terminal: false
});

process.stdin.setEncoding('utf8');
rl.once('line', () => {
	rl.on('line', readLine);
});

/**
 * Enter something like
 *
 * 5
 * 1 2 3 4 5
 *
 * @param line
 */
function readLine(line) {
	const numbers = line.toString().split(' ').map(Number);

	console.log(max_pairwise_product(numbers));
	process.exit();
}

function max_pairwise_product(numbers) {
	const length = numbers.length;

	if (length < 2) {
		throw new Error("Length is too short!");
	}

	if (length > 2 * Math.pow(10, 5)) {
		throw new Error("Length is too long!");
	}

	let biggestNumber = 0;
	let secondBiggestNumber = 0;
	if (numbers[0] >= numbers[1]) {
		biggestNumber = numbers[0];
		secondBiggestNumber = numbers[1];
	} else {
		biggestNumber = numbers[1];
		secondBiggestNumber = numbers[0];
	}

	for (let i = 2; i < length; i++) {
		if (numbers[i] >= biggestNumber) {
			secondBiggestNumber = biggestNumber;
			biggestNumber = numbers[i];
		} else if (numbers[i] >= secondBiggestNumber) {
			secondBiggestNumber = numbers[i];
		}
	}

	return biggestNumber * secondBiggestNumber;

}

module.exports = max_pairwise_product;

File: course1_algorithmic_toolbox/week1_programming_challenges/2_maximum_pairwise_product/MaxPairwiseProduct.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

import static java.lang.Math.pow;

public class MaxPairwiseProduct {

	/**
	 * Enter something like
	 *
	 * 5
	 * 1 2 3 4 5
	 *
	 * @param args
	 */
	public static void main(String[] args) {
		FastScanner scanner = new FastScanner(System.in);
		int n = scanner.nextInt();
		int[] numbers = new int[n];
		for (int i = 0; i < n; i++) {
			numbers[i] = scanner.nextInt();
		}
		System.out.println(getMaxPairwiseProduct(numbers));
	}

	private static long getMaxPairwiseProduct(int[] numbers) {
		int length = numbers.length;

		if (length < 2) {
			throw new RuntimeException("Length is too short!");
		}

		if (length > 2 * pow(10, 5)) {
			throw new RuntimeException("Length is too long!");
		}

		int biggestNumber = 0;
		int secondBiggestNumber = 0;
		if (numbers[0] >= numbers[1]) {
			biggestNumber = numbers[0];
			secondBiggestNumber = numbers[1];
		} else {
			biggestNumber = numbers[1];
			secondBiggestNumber = numbers[0];
		}

		for (int i = 2; i < length; i++) {
			if (numbers[i] >= biggestNumber) {
				secondBiggestNumber = biggestNumber;
				biggestNumber = numbers[i];
			} else if (numbers[i] >= secondBiggestNumber) {
				secondBiggestNumber = numbers[i];
			}
		}

		return (long) biggestNumber * (long) secondBiggestNumber;
	}

	static class FastScanner {
		BufferedReader br;
		StringTokenizer st;

		FastScanner(InputStream stream) {
			try {
				br = new BufferedReader(new
						InputStreamReader(stream));
			} catch (Exception e) {
				e.printStackTrace();
			}
		}

		String next() {
			while (st == null || !st.hasMoreTokens()) {
				try {
					st = new StringTokenizer(br.readLine());
				} catch (IOException e) {
					e.printStackTrace();
				}
			}
			return st.nextToken();
		}

		int nextInt() {
			return Integer.parseInt(next());
		}
	}

}
